File: my-folder/0146-lru-cache/solution.py
class Node:
    def __init__(self, key, val):
        self.val = val
        self.key = key
        self.prev = None
        self.next = None

class DLL:
    def __init__(self):
        self.head = None
        self.tail = None
    
    def insert(self, key, val):
        newNode = Node(key, val)

        if not self.head:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        
        return newNode
    
    def delete(self, node):
        if not node: return

        if not node.prev and not node.next:
            self.head = None

        elif node == self.head:
            self.head = self.head.next
            self.head.prev = None

        elif node == self.tail:
            node.prev.next = None
            self.tail = node.prev

        elif node.next and node.prev:
            node.prev.next = node.next
            node.next.prev = node.prev

        return node

class LRUCache:

    # @param capacity, an integer
    def __init__(self, capacity):
        self.capacity = capacity
        self.size = 0
        self.dll = DLL()
        self.cache = {}

    # @return an integer
    def get(self, key):
        if key in self.cache:
            self.dll.delete(self.cache[key])
            self.cache[key] = self.dll.insert(key, self.cache[key].val)
            return self.cache[key].val
        return -1

    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def put(self, key, value):
        if key in self.cache:
            self.dll.delete(self.cache[key])
            self.cache[key] = self.dll.insert(key, value)
            return 

        if self.size == self.capacity:
            del self.cache[self.dll.head.key]
            self.dll.delete(self.dll.head)
            self.size -= 1
        
        self.cache[key] = self.dll.insert(key, value)
        self.size += 1
        
        

File: my-folder/0146-lru-cache/solution.js
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.cache = new Map()
    this.dll = new DLL()
    this.capacity = capacity
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if (!this.cache.has(key)) {
        return -1
    }

    const cur = this.cache.get(key)
    this.dll.deleteNode(cur)
    this.dll.appendNode(cur)

    return cur.val
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    if (this.cache.has(key)) {
        this.dll.deleteNode(this.cache.get(key))
        this.cache.delete(key)
    }

    if (this.cache.size >= this.capacity) {
        const leastUsed = this.dll.head
        this.dll.deleteNode(leastUsed)
        this.cache.delete(leastUsed.key)
    }

    const newNode = this.dll.appendNode(new ListNode(value, key))
    this.cache.set(key, newNode)

};

class ListNode {
    constructor(val, key) {
        this.val = val
        this.key = key
        this.prev = null
        this.next = null
    }
}

class DLL {
    constructor() {
        this.head = null
        this.tail = null
    }
    
    appendNode(newNode) {
        if (!this.head) {
            this.head = newNode
            this.tail = newNode
        } else {
            this.tail.next = newNode
            newNode.prev = this.tail
            this.tail = newNode
        }
        return newNode
    }
    
    deleteNode(node) {
        if (!node) return
        
        if (node === this.head) {
            this.head = this.head.next
            if (this.head) {
                this.head.prev = null
            } else {
                this.tail = null
            }
        }
        else if (node === this.tail) {
            this.tail = this.tail.prev
            if (this.tail) {
                this.tail.next = null
            } else {
                this.head = null
            }
        }
        else {
            node.prev.next = node.next
            node.next.prev = node.prev
        }
    }
}

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
